equinox/nn/_attention.py

```diff
@@ -323,10 +438,53 @@ def __call__(
                 query_heads, key_heads, value_heads, key=keys
             )
         attn = attn.reshape(query_seq_length, -1)
+        out = jax.vmap(self.output_proj)(attn)
 
-        return jax.vmap(self.output_proj)(attn)
+        if state is sentinel:
+            return out
+        else:
+            return out, state
 
     def _project(self, proj, x):
         seq_length, _ = x.shape
         projection = jax.vmap(proj)(x)
         return projection.reshape(seq_length, self.num_heads, -1)
+
+
+def self_attention(
```

Comment:
I think this function was most useful back when we had MQA, and for which the `MultiheadAttention` API was a bit too much.

I don't think we're *quite* there yet (although it's definitely still pretty complicated), so I think I'd probably suggest removing this function for now.